app/javascript/packs/followButtons.js

```diff
@@ -0,0 +1,346 @@
+  url.search = searchParams;
+
+  fetch(url, {
+    method: 'GET',
+    headers: {
+      Accept: 'application/json',
+      'X-CSRF-Token': window.csrfToken,
+      'Content-Type': 'application/json',
+    },
+    credentials: 'same-origin',
+  })
+    .then((response) => response.json())
+    .then((idStatuses) => {
+      Object.keys(idStatuses).forEach((id) => {
+        idButtonHash[id].forEach((button) => {
+          updateInitialButtonUI(idStatuses[id], button);
+          button.dataset.buttonInitialized = true;
+        });
+      });
+    });
+}
+
+/**
+ * Sets up the initial state of all user follow buttons on the page,
+ * by obtaining the 'follow status' of each user and updating the associated buttons' UI.
+ */
+function initializeAllUserFollowButtons() {
+  const buttons = document.querySelectorAll(
+    '.follow-action-button.follow-user:not([data-fetched])',
+  );
+
+  if (buttons.length === 0) {
+    return;
+  }
+
+  const userIds = {};
+
+  Array.from(buttons, (button) => {
+    button.dataset.fetched = 'fetched';
+    const userStatus = document.body.getAttribute('data-user-status');
+
+    if (userStatus === 'logged-out') {
+      const { style } = JSON.parse(button.dataset.info);
+      addButtonFollowText(button, style);
+    } else {
+      const { id: userId } = JSON.parse(button.dataset.info);
+      if (userIds[userId]) {
+        userIds[userId].push(button);
+      } else {
+        userIds[userId] = [button];
+      }
+    }
+  });
+
+  if (Object.keys(userIds).length > 0) {
+    fetchUserFollowStatuses(userIds);
+  }
+}
+
+/**
+ * Individually fetches the current status of a follow button and updates the UI to match
+ *
+ * @param {HTMLElement} button
+ * @param {Object} buttonInfo The parsed buttonInfo object obtained from the button's data-attribute
+ */
+function fetchFollowButtonStatus(button, buttonInfo) {
+  button.dataset.fetched = 'fetched';
+
+  fetch(`/follows/${buttonInfo.id}?followable_type=${buttonInfo.className}`, {
+    method: 'GET',
+    headers: {
+      Accept: 'application/json',
+      'X-CSRF-Token': window.csrfToken,
+      'Content-Type': 'application/json',
+    },
+    credentials: 'same-origin',
+  })
+    .then((response) => response.text())
+    .then((followStatus) => {
+      updateInitialButtonUI(followStatus, button);
+      button.dataset.buttonInitialized = true;
+    });
+}
+
+/**
+ * Makes sure the initial state of follow buttons is fetched and presented in the UI.
+ * User follow buttons are initialized separately via bulk request
+ */
+function initializeNonUserFollowButtons() {
+  const nonUserFollowButtons = document.querySelectorAll(
+    '.follow-action-button:not(.follow-user)',
+  );
+
+  const user = userData();
+  nonUserFollowButtons.forEach((button) => {
+    const { info, fetched } = button.dataset;
+    const buttonInfo = JSON.parse(info);
+    if (
+      buttonInfo.className === 'Tag' &&
+      user &&
+      button.dataset.buttonInitialized !== 'true'
```

Comment:
Question (non-blocking): All the `data-button-initialized` attributes are to ensure that the buttons are in the correct state, e.g. Follow or Following?

Reply:
I think these were being used for testing purposes, but I'm pretty sure I can remove them now since we no longer rely on individual click listeners being attached. We have a separate `data-fetched` attribute that's used to make sure we don't refetch already known button info

Reply:
On a related note, I do have a data-attribute on the whole container to confirm when the click handler has been attached... I know we discussed moving away from this to disabling buttons but I'm wondering if we want to do that as part of this PR too 🤔

Reply:
In this case I think it makes sense not to as the event listener is on a container.

Reply:
I don't think so - the container is basically the whole inner page so if we make that a fieldset it would encompass every form field on the page 🙁

Reply:
Ahh, yeah, forgot it was that `<div />`. 🙃 We could disable all the follow buttons by default potentially, but maybe that is more headache than help. Worst case in the test, we check the data attribute on the container before we start clicking.

Reply:
Yeah I wonder if it is going to be an issue now... tests will wait until the button updates to e.g. 'Following' if that's what we're expecting, and the click handler should be available pretty much immediately 🤷‍♀️  One attribute on the container also doesn't seem as bad as peppering them around across all the buttons too

As always I'll re-run the specs a bunch locally and see how we're looking 😄